src/Components/Web/src/Forms/InputBase.cs

```diff
@@ -159,62 +160,53 @@ protected InputBase()
         protected abstract bool TryParseValueFromString(string? value, [MaybeNullWhen(false)] out TValue result, [NotNullWhen(false)] out string? validationErrorMessage);
 
         /// <summary>
-        /// Gets a string that indicates the status of the field being edited. This will include
-        /// some combination of "modified", "valid", or "invalid", depending on the status of the field.
-        /// </summary>
-        private string FieldClass
-            => EditContext.FieldCssClass(FieldIdentifier);
-
-        /// <summary>
-        /// Gets a CSS class string that combines the <c>class</c> attribute and <see cref="FieldClass"/>
-        /// properties. Derived components should typically use this value for the primary HTML element's
-        /// 'class' attribute.
+        /// Gets a CSS class string that combines the <c>class</c> attribute and and a string indicating
+        /// the status of the field being edited (a combination of "modified", "valid", and "invalid").
+        /// Derived components should typically use this value for the primary HTML element's 'class' attribute.
         /// </summary>
         protected string CssClass
         {
             get
             {
+                var fieldClass = EditContext?.FieldCssClass(FieldIdentifier);
+
                 if (AdditionalAttributes != null &&
                     AdditionalAttributes.TryGetValue("class", out var @class) &&
                     !string.IsNullOrEmpty(Convert.ToString(@class, CultureInfo.InvariantCulture)))
                 {
-                    return $"{@class} {FieldClass}";
+                    return string.IsNullOrEmpty(fieldClass) ? @class.ToString()! : $"{@class} {fieldClass}";
                 }
 
-                return FieldClass; // Never null or empty
```

Comment:
Since the `EditContext` is optional now, the class string can be empty if there is no class provided via `AdditionalAttributes`. This is the reason I added the extension method `RenderTreeBuilder.AddAttributeIfNotNullOrEmpty()`.

Reply:
This is super unrelated to your change, but could we capture the result of `Convert.ToString(@class, CultureInfo.InvariantCulture)` and use that inside the `if`? It feels like that's what we should have been using instead of relying on `@class.ToString()` to produce the right representation inside the interpolated string.

Reply:
@pranavkm I've made this change. Since `InputRadio` had nearly identical logic for creating its class string, I've added a static class `AttributeUtilities` containing a method `CombineClassNames` that does this (`InputRadio` cannot inherit the implementation in `InputBase` because it does not extend `InputBase`).